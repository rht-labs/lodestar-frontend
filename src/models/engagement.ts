export interface Engagement {
  name: string;
}
export class Engagement {
  constructor({ name }: Engagement) {
    this.name = name;
  }
  name: string;
}
