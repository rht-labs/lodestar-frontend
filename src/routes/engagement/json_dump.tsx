import React, { useEffect, useState } from 'react';
import { DataCard } from '../../components/engagement_data_cards/data_card';
import AceEditor from 'react-ace';
import 'ace-builds/src-noconflict/mode-json';
import {
  Alert,
  AlertVariant,
  Button,
  Flex,
  FlexItem,
  PageSection,
  PageSectionVariants,
} from '@patternfly/react-core';
import 'ace-builds/src-noconflict/theme-github';
import { Link, useRouteMatch } from 'react-router-dom';
import { Feature } from '../../components/feature/feature';

export interface EngagementJsonDumpProps {
  json: string;
  onSave(json: string): void;
}
export function EngagementJsonDump(props: EngagementJsonDumpProps) {
  const [currentValue, setCurrentValue] = useState<string>(props.json);
  useEffect(() => setCurrentValue(props.json), [props.json]);
  const handleChange = (value: string) => setCurrentValue(value);
  const handleSave = () => props.onSave(currentValue);

  const { url } = useRouteMatch();

  return (
    <DataCard title="Engagement JSON">
      <Alert
        isInline
        title={
          <span>
            Only use this tool if you know what you are doing. If you came here
            by mistake,{' '}
            <Link to={`${url.replace('/json', '')}`}>click here</Link> to return
            to the engagement page.
          </span>
        }
        variant={AlertVariant.danger}
      ></Alert>
      <CodeEditor onChange={handleChange} value={currentValue} />
      <PageSection variant={PageSectionVariants.light}>
        <Flex justifyContent={{ default: 'justifyContentFlexEnd' }}>
          <FlexItem>
            <Feature name="admin">
              <Button onClick={handleSave}>Save</Button>
            </Feature>
          </FlexItem>
        </Flex>
      </PageSection>
    </DataCard>
  );
}

interface CodeEditorProps {
  value: string;
  onChange(value: string): void;
}
const CodeEditor = (props: CodeEditorProps) => {
  return (
    <AceEditor
      width={'100%'}
      mode={'json'}
      value={props.value}
      onChange={props.onChange}
      theme="github"
    />
  );
};
