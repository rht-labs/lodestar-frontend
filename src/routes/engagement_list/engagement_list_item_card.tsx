import React from 'react';
import {
  Button,
  Card,
  CardBody,
  Title,
  Grid, GridItem,
} from '@patternfly/react-core';
import { useHistory } from 'react-router';

export interface DataCardProps {
  title: string;
  customer: string;
  project: string;
  children: any;
}

export function EngagementListItemCard(props: DataCardProps) {
  const url = `/app/engagements/${props.customer}/${props.project}`;
  const history = useHistory();

  return (
    <>
      <Card isCompact style={{ margin: '1rem', padding: '2rem 0.5rem' }}>
        <CardBody style={{ padding: '0 2rem' }}>
          <Grid hasGutter style={{alignItems: 'center'}}>
            <GridItem md={12} lg={3}>
              <Button variant="link"
                      isInline
                      style={ {whiteSpace: 'normal', wordWrap: 'break-word', textAlign:'left'}}
                      onClick={() => { history.push(url)} }>
                <Title headingLevel="h3" style={{ fontWeight: 'bolder'}}>
                  {props.title}
                </Title>
              </Button>
              <Title style={{ fontWeight: 'normal'}} headingLevel="h4">
                {props.customer}
              </Title>
            </GridItem>
            <GridItem md={12} lg={7}>
              {props.children}
            </GridItem>
            <GridItem sm={12} md={4} lg={2}>
              <Button
                onClick={() => {
                  history.push(url);
                }}
                variant="secondary"
                data-cy={'view_engagement_button'}
              >
                View Engagement
              </Button>
            </GridItem>
          </Grid>
        </CardBody>
      </Card>
    </>
  );
}
