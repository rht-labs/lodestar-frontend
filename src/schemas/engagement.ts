import faker from 'faker/locale/en_US';
import { LaunchData } from './launch_data';
import { GitCommit } from './git_commit';
import { CreationDetails } from './creation_details';
import { ClusterStatus } from './cluster_status';
export interface Engagement {
  additional_details?: string;
  archive_date: Date;
  commits: GitCommit[];
  customer_contact_email: string;
  customer_contact_name: string;
  customer_name: string;
  description: string;
  end_date: Date;
  engagement_lead_email: string;
  engagement_lead_name: string;
  engagement_users: Array<any>;
  last_update: string;
  location: string;
  mongo_id?: string;
  ocp_cloud_provider_name: string;
  ocp_cloud_provider_region: string;
  ocp_cluster_size: string;
  ocp_persistent_storage_size: string;
  ocp_sub_domain: string;
  ocp_version: string;
  project_id: number;
  project_name: string;
  start_date: Date;
  technical_lead_email: string;
  technical_lead_name: string;
  launch?: LaunchData;
  creation_details: CreationDetails;
  last_update_by_name: string;
  suggested_subdomain?: string;
  status: ClusterStatus;
}
export abstract class Engagement {
  static fromFake(staticData = false): Engagement {
    return {
      additional_details: staticData
        ? 'Additional information here'
        : faker.lorem.paragraphs(2),
      archive_date: staticData ? new Date(2020, 6, 30) : faker.date.recent(),
      commits: [],
      customer_contact_email: staticData
        ? 'bob@doe.com'
        : faker.internet.email(),
      customer_contact_name: staticData
        ? 'Bob Doe'
        : `${faker.name.firstName()} ${faker.name.lastName()}`,
      customer_name: staticData ? 'NASA' : faker.company.companyName(),
      description: staticData
        ? "It's rocket science"
        : faker.lorem.paragraphs(5),
      end_date: staticData ? new Date(2020, 6, 1) : faker.date.future(),
      engagement_users: staticData ? [] : [],
      engagement_lead_email: staticData
        ? 'alice@doe.com'
        : faker.internet.email(),
      engagement_lead_name: staticData
        ? 'Alice Doe'
        : `${faker.name.firstName()} ${faker.name.lastName()}`,
      last_update: staticData ? '2020-01-01' : new Date().toString(),
      location: staticData
        ? 'Nashville, TN'
        : `${faker.address.city()}, ${faker.address.stateAbbr()}`,
      mongo_id: '1',
      ocp_cloud_provider_name: staticData ? 'AWS' : 'AWS',
      ocp_cloud_provider_region: staticData
        ? 'N. Virginia'
        : faker.lorem.word(),
      ocp_cluster_size: staticData ? 'Large' : 'Large',
      ocp_persistent_storage_size: staticData
        ? '100G'
        : `${faker.random.number()}G`,
      ocp_sub_domain: staticData ? 'ordoabchao' : faker.lorem.word(),
      ocp_version: staticData ? '4.4' : faker.random.number().toString(),
      project_id: staticData ? 1 : faker.random.number(),
      project_name: staticData ? 'Boots on the Moon' : faker.company.bsNoun(),
      start_date: staticData ? new Date(2020, 5, 1) : faker.date.recent(),
      technical_lead_email: staticData ? 'eve@doe.com' : faker.internet.email(),
      technical_lead_name: staticData
        ? 'Eve Doe'
        : `${faker.name.firstName()} ${faker.name.lastName()}`,
      creation_details: staticData
        ? {
            created_by_email: 'dwasinger@redhat.com',
            created_by_user: 'dwasinge',
            created_on: new Date(2020, 1, 1),
          }
        : {
            created_by_email: faker.internet.email(),
            created_by_user: `${faker.name.firstName()} ${faker.name.lastName()}`,
            created_on: faker.date.recent(),
          },
      last_update_by_name: staticData
        ? 'James Doe'
        : `${faker.name.firstName()} ${faker.name.lastName()}`,
      launch: staticData
        ? null
        : faker.random.boolean()
        ? {
            launched_by: faker.name.firstName(),
            launched_date_time: faker.date.recent(),
          }
        : null,
      status: ClusterStatus.fromFake(staticData),
    };
  }
}

export enum EngagementStatus {
  active = 'active',
  past = 'past',
  upcoming = 'upcoming',
}

export const getEngagementStatus = (
  engagement: Engagement
): EngagementStatus => {
  if (!engagement) {
    return null;
  }
  const Today = new Date();
  const { launch, end_date } = engagement;
  const hasLaunched = !!launch?.launched_date_time;
  if (hasLaunched && end_date >= Today) {
    return EngagementStatus.active;
  } else if (hasLaunched && end_date < Today) {
    return EngagementStatus.past;
  } else {
    return EngagementStatus.upcoming;
  }
};
