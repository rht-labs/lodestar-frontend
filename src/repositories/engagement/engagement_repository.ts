import { Engagement } from '../../models/engagement';

export abstract class EngagementRepository {
  abstract async fetchEngagements(): Promise<Engagement[]>;
}
