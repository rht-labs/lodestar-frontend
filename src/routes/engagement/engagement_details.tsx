import React, { useEffect } from 'react';
import { Engagement } from '../../schemas/engagement';
import { useEngagement } from '../../context/engagement_context/engagement_hook';
import { Route, useParams, Switch, useRouteMatch } from 'react-router';
import { getValidatorsFromEngagementFormConfig } from '../../common/config_validator_adapter';
import {
  ValidationProvider,
  FormValidator,
} from '../../context/validation_context/validation_context';
import { EngagementDetailsViewTemplate } from '../../layout/engagement_details_view';
import { EngagementOverview } from './overview';
import { EngagementJsonDump } from './json_dump';
import { EngagementJsonSerializer } from '../../serializers/engagement/engagement_json_serializer';

export interface EngagementViewProps {
  currentEngagement?: Engagement;
}

export interface EngagementDetailViewProps {
  error: Error;
  currentEngagement: Engagement;
  createEngagementPoll: (
    engagement: Engagement
  ) => Promise<{ cancel: () => void }>;
  getEngagement: (
    customer_name: string,
    project_name: string
  ) => Promise<Engagement>;
}
export const validatorsWithDateValidators = (
  validators: FormValidator,
  dates: Pick<Engagement, 'start_date' | 'end_date'>
): FormValidator => {
  return {
    ...validators,
    end_date: [
      ...(validators['end_date'] ?? []),
      endDate => {
        const result =
          dates?.start_date > endDate
            ? 'End date must be after the start date'
            : null;
        return result;
      },
    ],
    archive_date: [
      ...(validators['archive_date'] ?? []),
      archiveDate => {
        const result =
          archiveDate < dates?.end_date
            ? 'Archive date must be after the end date'
            : null;
        return result;
      },
    ],
  };
};

export const EngagementDetailView = () => {
  const { project_name, customer_name } = useParams<{
    customer_name: string;
    project_name: string;
  }>();

  const {
    createEngagementPoll,
    saveEngagement,
    currentEngagement,
    currentChanges,
    engagementFormConfig,
  } = useEngagement({ projectName: project_name, customerName: customer_name });

  useEffect(() => {
    let engagementPoll;
    if (currentEngagement?.project_name && currentEngagement?.customer_name) {
      engagementPoll = createEngagementPoll(currentEngagement);
    }
    return async () => {
      (await engagementPoll)?.cancel?.();
    };
  }, [currentEngagement, createEngagementPoll]);

  const validators = getValidatorsFromEngagementFormConfig(
    engagementFormConfig
  );
  const { url } = useRouteMatch();
  const serializer = new EngagementJsonSerializer();
  const {
    commits,
    uuid,
    project_id,
    mongo_id,
    creation_details,
    status,
    launch,
    ...editableFields
  } = currentEngagement ?? {};
  const { end_date, start_date } = currentChanges;

  return (
    <ValidationProvider
      validators={validatorsWithDateValidators(validators, {
        end_date,
        start_date,
      })}
    >
      <EngagementDetailsViewTemplate
        engagement={currentEngagement}
        onSave={saveEngagement}
      >
        <Switch>
          <Route path={`${url}/json`}>
            <EngagementJsonDump
              json={JSON.stringify(
                serializer.serialize((editableFields ?? {}) as Engagement),
                null,
                2
              )}
              onSave={value => {
                saveEngagement(serializer.deserialize(JSON.parse(value)));
              }}
            />
          </Route>
          <Route path={`${url}/`}>
            <EngagementOverview />
          </Route>
        </Switch>
      </EngagementDetailsViewTemplate>
    </ValidationProvider>
  );
};
