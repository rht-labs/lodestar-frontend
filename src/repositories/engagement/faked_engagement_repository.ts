import { EngagementRepository } from './engagement_repository';
import { Engagement } from '../../models/engagement';
import faker from 'faker';

export class FakedEngagementRepository extends EngagementRepository {
  async fetchEngagements(): Promise<Engagement[]> {
    return new Array(8)
      .fill(null)
      .map(() => new Engagement({ name: faker.company.companyName() }));
  }
}
