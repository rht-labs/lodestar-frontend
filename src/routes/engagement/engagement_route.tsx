import React from 'react';
import { useAnalytics } from '../../context/analytics_context/analytics_context';
import { useSession } from '../../context/auth_context/auth_context';
import { useEngagementFormConfig } from '../../context/engagement_config_context/engagement_config_hook';
import { EngagementProvider } from '../../context/engagement_context/engagement_context';
import { useFeedback } from '../../context/feedback_context/feedback_context';
import { useServiceProviders } from '../../context/service_provider_context/service_provider_context';
import ScrollToTop from '../scroll_to_top';
import { EngagementDetailView } from './engagement_details';

export const EngagementRoute = () => {
  const { engagementService, categoryService } = useServiceProviders();
  const feedback = useFeedback();
  const analyticsContext = useAnalytics();
  const authContext = useSession();
  const { engagementFormConfig } = useEngagementFormConfig(engagementService);
  return (
    <EngagementProvider
      authContext={authContext}
      feedbackContext={feedback}
      analyticsContext={analyticsContext}
      categoryService={categoryService}
      engagementService={engagementService}
      engagementFormConfig={engagementFormConfig}
    >
      <ScrollToTop>
        <EngagementDetailView />
      </ScrollToTop>
    </EngagementProvider>
  );
};
