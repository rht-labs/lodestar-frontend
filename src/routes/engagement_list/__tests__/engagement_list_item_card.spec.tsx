import React from 'react';
import { cleanup, render } from '@testing-library/react';
import { MemoryRouter } from 'react-router';
import '@testing-library/jest-dom/extend-expect';
import { EngagementListItemCard } from '../engagement_list_item_card';
import { Engagement } from '../../../schemas/engagement';

afterEach(cleanup);

describe('Engagement Card', () => {
  test('renders the components', () => {
    const engagement = Engagement.fromFake(true);
    const { getByText, getByTestId } = render(
      <MemoryRouter>
        <EngagementListItemCard
          customer={engagement.customer_name}
          project={engagement.project_name}
          title={'Test Title'}
        >
          <div data-testid="childComponent"> Test Child Component </div>
        </EngagementListItemCard>
      </MemoryRouter>
    );
    expect(getByTestId('childComponent')).toBeInTheDocument();
    expect(getByText('View Engagement')).toBeInTheDocument();
    expect(getByText('Test Title')).toBeInTheDocument();
  });

  test('matches the snapshot', () => {
    const engagement = Engagement.fromFake(true);
    expect(
      render(
        <MemoryRouter>
          <EngagementListItemCard
            customer={engagement.customer_name}
            project={engagement.project_name}
            title={'Test Title'}
          >
            <div> Test Child Component </div>
          </EngagementListItemCard>
        </MemoryRouter>
      )
    ).toMatchSnapshot();
  });
});
