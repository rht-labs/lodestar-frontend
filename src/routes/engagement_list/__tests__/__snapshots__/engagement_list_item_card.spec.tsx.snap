// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Engagement Card matches the snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <article
        class="pf-c-card pf-m-compact"
        style="margin: 1rem; padding: 2rem 0.5rem;"
      >
        <div
          class="pf-c-card__body"
          style="padding: 0px 2rem;"
        >
          <div
            class="pf-l-grid pf-m-gutter"
            style="align-items: center;"
          >
            <div
              class="pf-l-grid__item pf-m-12-col-on-md pf-m-3-col-on-lg"
            >
              <button
                class="pf-c-button pf-m-link pf-m-inline"
                data-ouia-component-type="PF4/Button"
                data-ouia-safe="true"
                style="white-space: normal; word-wrap: break-word; text-align: left;"
                type="button"
              >
                <h3
                  class="pf-c-title pf-m-lg"
                  style="font-weight: bolder;"
                >
                  Test Title
                </h3>
              </button>
              <h4
                class="pf-c-title pf-m-md"
                style="font-weight: normal;"
              />
            </div>
            <div
              class="pf-l-grid__item pf-m-12-col-on-md pf-m-7-col-on-lg"
            >
              <div>
                 Test Child Component 
              </div>
            </div>
            <div
              class="pf-l-grid__item pf-m-12-col-on-sm pf-m-4-col-on-md pf-m-2-col-on-lg"
            >
              <button
                class="pf-c-button pf-m-secondary"
                data-cy="view_engagement_button"
                data-ouia-component-type="PF4/Button"
                data-ouia-safe="true"
                type="button"
              >
                View Engagement
              </button>
            </div>
          </div>
        </div>
      </article>
    </div>
  </body>,
  "container": <div>
    <article
      class="pf-c-card pf-m-compact"
      style="margin: 1rem; padding: 2rem 0.5rem;"
    >
      <div
        class="pf-c-card__body"
        style="padding: 0px 2rem;"
      >
        <div
          class="pf-l-grid pf-m-gutter"
          style="align-items: center;"
        >
          <div
            class="pf-l-grid__item pf-m-12-col-on-md pf-m-3-col-on-lg"
          >
            <button
              class="pf-c-button pf-m-link pf-m-inline"
              data-ouia-component-type="PF4/Button"
              data-ouia-safe="true"
              style="white-space: normal; word-wrap: break-word; text-align: left;"
              type="button"
            >
              <h3
                class="pf-c-title pf-m-lg"
                style="font-weight: bolder;"
              >
                Test Title
              </h3>
            </button>
            <h4
              class="pf-c-title pf-m-md"
              style="font-weight: normal;"
            />
          </div>
          <div
            class="pf-l-grid__item pf-m-12-col-on-md pf-m-7-col-on-lg"
          >
            <div>
               Test Child Component 
            </div>
          </div>
          <div
            class="pf-l-grid__item pf-m-12-col-on-sm pf-m-4-col-on-md pf-m-2-col-on-lg"
          >
            <button
              class="pf-c-button pf-m-secondary"
              data-cy="view_engagement_button"
              data-ouia-component-type="PF4/Button"
              data-ouia-safe="true"
              type="button"
            >
              View Engagement
            </button>
          </div>
        </div>
      </div>
    </article>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
